EgammaAnalysis/ElectronTools/src/ElectronEnergyCalibratorRun2.cc

```diff
@@ -31,63 +31,145 @@ void ElectronEnergyCalibratorRun2::initPrivateRng(TRandom *rnd)
-  float scale_gain = _correctionRetriever.ScaleCorrectionUncertainty(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 4);
-  std::vector<float> scale_uncs;
+  float scale_stat_up = scale + correctionRetriever_.ScaleCorrectionUncertainty(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 1);
+  float scale_stat_dn = scale - correctionRetriever_.ScaleCorrectionUncertainty(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 1);
+  float scale_syst_up = scale + correctionRetriever_.ScaleCorrectionUncertainty(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 2);
+  float scale_syst_dn = scale - correctionRetriever_.ScaleCorrectionUncertainty(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 2);
+  float scale_gain_up = scale + correctionRetriever_.ScaleCorrectionUncertainty(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 4);
+  float scale_gain_dn = scale - correctionRetriever_.ScaleCorrectionUncertainty(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 4);
+  float resol_rho_up  = correctionRetriever_.getSmearingSigma(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 1., 0.);
+  float resol_rho_dn  = correctionRetriever_.getSmearingSigma(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, -1., 0.);
+  float resol_phi_up  = correctionRetriever_.getSmearingSigma(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 0., 1.);
+  float resol_phi_dn  = correctionRetriever_.getSmearingSigma(runNumber, electron.isEB(), electron.full5x5_r9(), aeta, et, gainSeedSC, 0., -1.);
+  std::vector<float> uncertainties;
 
-  double newEcalEnergy, newEcalEnergyError;
+  double newEcalEnergy = electron.correctedEcalEnergy();
+  double newEcalEnergyError = electron.correctedEcalEnergyError();
   std::pair<float, float> combinedMomentum;
 
-  if (isMC_) {
-    double corr = 1.0 + smear * gauss(id);
-    newEcalEnergy      = electron.correctedEcalEnergy() * corr;
-    newEcalEnergyError = std::hypot(electron.correctedEcalEnergyError() * corr, smear * newEcalEnergy);
+  if ((eventIsMC < 0 && isMC_) || (eventIsMC == 1)) {
+    double rndm = gauss(id);
+    double corr = 1.0 + smear * rndm;
+    double corr_rho_up = 1.0 + resol_rho_up * rndm;
+    double corr_rho_dn = 1.0 + resol_rho_dn * rndm;
+    double corr_phi_up = 1.0 + resol_phi_up * rndm;
+    double corr_phi_dn = 1.0 + resol_phi_dn * rndm;
 
+    electron.setCorrectedEcalEnergy(newEcalEnergy * corr);
+    electron.setCorrectedEcalEnergyError(std::hypot(newEcalEnergyError * corr, smear * newEcalEnergy * corr));
+    combinedMomentum = epCombinationTool_->combine(electron);
+    uncertainties.push_back(combinedMomentum.first);
+    uncertainties.push_back(combinedMomentum.first);
```

Comment:
Because in MC the correction does not depend on the scale, only on the resolution. So there is no energy scale uncertainty. I just copy the central value 6 times to keep the order of the vector.